File: Tests/AINotebookCoreTests/NoteVersionStoreTests.swift
import XCTest
@testable import AINotebookCore

@MainActor
final class NoteVersionStoreTests: XCTestCase {

    func testUpdateSnapshotsPreviousBody() throws {
        let store = try NotebookStore(path: .inMemory)
        let nb = try store.createNotebook(name: "NB")
        let n = try store.createNote(notebookId: nb.id!, title: "T", bodyMd: "v1")
        try store.updateNote(id: n.id!, title: "T", bodyMd: "v2")
        let versions = try store.noteVersions(noteId: n.id!)
        XCTAssertEqual(versions.count, 1)
        XCTAssertEqual(versions[0].bodyMd, "v1")
        XCTAssertEqual(versions[0].reason, .autosave)
    }

    func testManualSnapshot() throws {
        let store = try NotebookStore(path: .inMemory)
        let nb = try store.createNotebook(name: "NB")
        let n = try store.createNote(notebookId: nb.id!, title: "T", bodyMd: "v1")
        try store.snapshotNoteVersion(noteId: n.id!, reason: .manual)
        let versions = try store.noteVersions(noteId: n.id!)
        XCTAssertEqual(versions.count, 1)
        XCTAssertEqual(versions[0].bodyMd, "v1")
        XCTAssertEqual(versions[0].reason, .manual)
    }

    func testRestoreCreatesNewSnapshotAndOverwritesBody() throws {
        let store = try NotebookStore(path: .inMemory)
        let nb = try store.createNotebook(name: "NB")
        let n = try store.createNote(notebookId: nb.id!, title: "T", bodyMd: "v1")
        try store.updateNote(id: n.id!, title: "T", bodyMd: "v2")
        let versions = try store.noteVersions(noteId: n.id!)
        let v1 = try XCTUnwrap(versions.first)
        try store.restoreNoteVersion(versionId: v1.id!)
        let reloaded = try XCTUnwrap(try store.note(id: n.id!))
        XCTAssertEqual(reloaded.bodyMd, "v1")
        let all = try store.noteVersions(noteId: n.id!)
        XCTAssertGreaterThanOrEqual(all.count, 2)
    }

    func testFiftyRowCapPrunesOldest() throws {
        let store = try NotebookStore(path: .inMemory)
        let nb = try store.createNotebook(name: "NB")
        let n = try store.createNote(notebookId: nb.id!, title: "T", bodyMd: "v0")
        for i in 1...60 {
            try store.updateNote(id: n.id!, title: "T", bodyMd: "v\(i)")
        }
        let versions = try store.noteVersions(noteId: n.id!)
        XCTAssertLessThanOrEqual(versions.count, 50)
    }
}

File: windows/tests/AINotebook.Core.Tests/Storage/NoteVersionStoreTests.cs
using AINotebook.Core.Models;
using AINotebook.Core.Storage;
using Xunit;

namespace AINotebook.Core.Tests.Storage;

public class NoteVersionStoreTests
{
    private static (NotebookStore store, long noteId) Fresh()
    {
        var store = new NotebookStore(StorePath.InMemory);
        var nb = store.CreateNotebook("N").Id!.Value;
        var note = store.CreateNote(nb, "t", "v1");
        return (store, note.Id!.Value);
    }

    [Fact]
    public void UpdateNoteSnapshotsPreviousBodyAsAutosave()
    {
        var (store, noteId) = Fresh();
        using (store)
        {
            store.UpdateNote(noteId, "t", "v2");
            var versions = store.NoteVersions(noteId);
            Assert.Single(versions);
            Assert.Equal(NoteVersionReason.Autosave, versions[0].Reason);
            Assert.Equal("v1", versions[0].BodyMd);
        }
    }

    [Fact]
    public void ManualSnapshotUsesManualReason()
    {
        var (store, noteId) = Fresh();
        using (store)
        {
            store.SnapshotNoteVersion(noteId, NoteVersionReason.Manual);
            Assert.Equal(NoteVersionReason.Manual, store.NoteVersions(noteId).Single().Reason);
        }
    }

    [Fact]
    public void RestoreSnapshotsCurrentAsRestoreThenOverwritesBody()
    {
        var (store, noteId) = Fresh();
        using (store)
        {
            // first update creates an autosave version holding "v1"
            store.UpdateNote(noteId, "t", "v2");
            var v1Version = store.NoteVersions(noteId).Single(v => v.BodyMd == "v1");
            store.RestoreNoteVersion(v1Version.Id!.Value);
            Assert.Equal("v1", store.Note(noteId)!.BodyMd);
            var versions = store.NoteVersions(noteId);
            Assert.True(versions.Count >= 2);
            Assert.Contains(versions, v => v.Reason == NoteVersionReason.Restore && v.BodyMd == "v2");
        }
    }

    [Fact]
    public void VersionsCappedAtFiftyOldestPruned()
    {
        var (store, noteId) = Fresh();
        using (store)
        {
            for (int i = 0; i < 60; i++)
                store.UpdateNote(noteId, "t", $"body {i}");
            Assert.True(store.NoteVersions(noteId).Count <= 50);
        }
    }
}
